pkg/controller/v1alpha2/experiment/experiment_controller.go

```diff
@@ -80,17 +108,251 @@ type ReconcileExperiment struct {
 		}
 		// Error reading the object - requeue the request.
+		logger.Error(err, "Experiment Get error")
 		return reconcile.Result{}, err
 	}
+	instance := original.DeepCopy()
+
+	if instance.IsCompleted() {
+
+		return reconcile.Result{}, nil
+
+	}
+	if !instance.IsCreated() {
+		//Experiment not created in DB
+		err = util.CreateExperimentInDB(instance)
+		if err != nil {
+			logger.Error(err, "Create experiment in DB error")
+			return reconcile.Result{}, err
+		}
+
+		if instance.Status.StartTime == nil {
+			now := metav1.Now()
+			instance.Status.StartTime = &now
+		}
+		msg := "Experiment is created"
+		instance.MarkExperimentStatusCreated(util.ExperimentCreatedReason, msg)
+		requeue = true
+	} else {
+		// Experiment already created in DB
+		err := r.ReconcileExperiment(instance)
+		if err != nil {
+			logger.Error(err, "Reconcile experiment error")
+			return reconcile.Result{}, err
+		}
+	}
+
+	if !equality.Semantic.DeepEqual(original.Status, instance.Status) {
+		//assuming that only status change
+		err = util.UpdateExperimentStatusInDB(instance)
+		if err != nil {
+			logger.Error(err, "Update experiment status in DB error")
+			return reconcile.Result{}, err
+		}
+		err = r.Status().Update(context.TODO(), instance)
+		if err != nil {
+			logger.Error(err, "Update experiment instance status error")
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{Requeue: requeue}, nil
+}
+
+func (r *ReconcileExperiment) ReconcileExperiment(instance *experimentsv1alpha2.Experiment) error {
+
+	var err error
+	logger := log.WithValues("Experiment", types.NamespacedName{Name: instance.GetName(), Namespace: instance.GetNamespace()})
+	trials := &trialsv1alpha2.TrialList{}
+	labels := map[string]string{"experiment": instance.Name}
+	lo := &client.ListOptions{}
+	lo.MatchingLabels(labels).InNamespace(instance.Namespace)
+
+	err = r.List(context.TODO(), lo, trials)
+	if err != nil {
+		logger.Error(err, "Trial List error")
+		return err
+	}
+	if len(trials.Items) > 0 {
+		err := util.UpdateExperimentStatus(instance, trials)
+		if err != nil {
+			logger.Error(err, "Update experiment status error")
+			return err
+		}
+	}
+	reconcileRequired := !instance.IsCompleted()
+	if reconcileRequired {
+		r.ReconcileTrials(instance)
+	}
+	return err
+}
+
+func (r *ReconcileExperiment) ReconcileTrials(instance *experimentsv1alpha2.Experiment) error {
+
+	var err error
+	logger := log.WithValues("Experiment", types.NamespacedName{Name: instance.GetName(), Namespace: instance.GetNamespace()})
+	parallelCount := 0
+
+	if instance.Spec.ParallelTrialCount != nil {
+		parallelCount = *instance.Spec.ParallelTrialCount
+	} else {
+		parallelCount = 3
+	}
+	activeCount := instance.Status.TrialsPending + instance.Status.TrialsRunning
+	completedCount := instance.Status.TrialsSucceeded + instance.Status.TrialsFailed + instance.Status.TrialsKilled
+
+	if activeCount > parallelCount {
+		deleteCount := activeCount - parallelCount
```

Comment:
I think line 215 and 223 can change to `if  activeCount >= parallelCount {} else {}`